### ISSUE TITLE
CUDA decorator/compiler argument to set literal precision

### ISSUE THREAD

**Issue by uchytilc** [numba/numba#4676]
I think this has been raised a few times but usually its in the context of converting every float to either a float32 or float64. If you want to do 32 bit operations (or anything other than 64 bit) with literals the current work around is to add np.TYPE32 to all literals in order to do 32 bit operations. The fact that literals almost always cause a cast to 64 often slows down kernels and does so in a very non-obvious way (I have seen a number of sample code snippets posted online where users add float64/int64 literals to float32/int32 inputs and don't realize that this is slowing down their code). It would be helpful if, in the cuda.jit decorator (or somewhere else perhaps), we could specify the precision of all literals in the kernel (the default can be 64 bit to maintain consistency with current code). I don't know how the python -> ptx is done but I would imagine that just setting all literals to a specific precision wouldn't be that hard (if someone can show me in Numba where this could be changed I can give it a try). If the default precision of a literal is changed, it also shouldn't cause any major problems (as far as I can tell) as, if an operation is performed on input data with the literal, it will just be cast up if necessary anyway.

**Comment by gmarkall** [comment]
I agree that this is a performance issue in a lot of cases. It's on my radar to look at (though not necessarily something I'll get to in the near future).
